sdk/search/azure-search-documents/azure/search/documents/_internal/_search_index_document_batching_client.py

```diff
@@ -0,0 +1,280 @@
+    def succeeded_actions(self):
+        # type: () -> List[IndexAction]
+        """The list of currently succeeded index actions in queue.
+
+        :rtype: List[IndexAction]
+        """
+        return self._index_documents_batch.succeeded_actions
+
+    @property
+    def failed_actions(self):
+        # type: () -> List[IndexAction]
+        """The list of currently failed index actions in queue.
+
+        :rtype: List[IndexAction]
+        """
+        return self._index_documents_batch.failed_actions
+
+    @property
+    def batch_size(self):
+        # type: () -> int
+        return self._batch_size
+
+    def close(self):
+        # type: () -> None
+        """Close the :class:`~azure.search.SearchClient` session.
+
+        """
+        return self._client.close()
+
+    def flush(self, raise_error=False):
+        # type: (bool) -> None
+        """Flush the batch.
+
+        :param bool raise_error: raise error if there are failures during flushing
+            Default to False which re-queue the failed tasks and retry on next flush.
+        """
+        # get actions
+        actions = self._index_documents_batch.dequeue_actions()
+        try:
+            results = self._index_documents_actions(actions=actions)
+            # re-queue 207:
+            if not self._index_key:
+                client = SearchServiceClient(self._endpoint)
+                kwargs = {"headers": self._merge_client_headers({})}
+                result = client.indexes.get(self._index_name, **kwargs)
+                if not result:
+                    # Cannot find the index
+                    self._index_key = ""
+                else:
+                    for field in result.fields:
+                        if field.key:
+                            self._index_key = field.name
+                            break
+
+            has_error = False
+
+            for result in results:
+                action = [x for x in actions if x.get(self._index_key) == result.key]
+                if is_retryable_status_code(result.status_code):
+                    self._index_documents_batch.enqueue_actions(action)
+                    has_error = True
+                elif result.status_code in [200, 201]:
+                    if self._persistence:
+                        self._persistence.remove_queued_actions(action)
+                        self._persistence.add_succeeded_actions(action)
+                    self._index_documents_batch.enqueue_succeeded_actions(action)
+                else:
+                    if self._persistence:
+                        self._persistence.remove_queued_actions(action)
+                        self._persistence.add_failed_actions(action)
+                    self._index_documents_batch.enqueue_failed_actions(action)
+                    has_error = True
+
+            if has_error and raise_error:
+                raise HttpResponseError(message="Some actions failed. Failed actions are re-queued.")
```

Comment:
What should the customer do? It seems it has been requeued for him, so is there any actions he should take or just wait more?

Reply:
This error only raise when user calls flush and explicitly set raise_error to True.

This is to relieve the shutdown/destructors lose buffered message issue.

In our design, user needs to call flush and handle re-queued tasks (e.g. dump into a storage to save the state) before they close/destruct the batch client.